sw/device/tests/sim_dv/sleep_pin_mio_dio_val_test.c

```diff
@@ -0,0 +1,195 @@
+
+static uint8_t kMioPads[NUM_MIO_PADS] = {0};
+static uint8_t kDioPads[NUM_DIO_PADS] = {0};
+
+/** Configure pinmux retention value.
+ *
+ * Each gen32 can cover 16 PADs randomization. When each pad ret val draws
+ * **3**, the code calls gen32_range to choose from 0 to 2.
+ */
+void draw_pinmux_ret(const uint32_t num_pins, uint8_t *arr) {
+  for (int i = 0; i < num_pins; i += 16) {
+    uint32_t val = rand_testutils_gen32();
+    uint32_t min_idx = (i + 16 < num_pins) ? i + 16 : num_pins;
+
+    for (int j = i; j < min_idx; j++) {
+      /* bit slice 2b at a time and if it is 3, redraw */
+      arr[j] = (val >> ((j & 0xF) * 2)) & 0x3;
+      if (arr[j] == 3) {
+        arr[j] = (uint8_t)rand_testutils_gen32_range(0, 2);
+      }
+    }
+  }
+}
+
+/** Send the chosen values to UVM tb.
+ *
+ * Format:
+ *   Each log sends up to 16 PADs values.
+ *
+ *   "DIO [i]: {}"
+ *
+ *   "MIO [i]: {}"
+ */
+void print_chosen_values(void) {
+  LOG_INFO("BEGIN Chosen Retention Types");
+  for (int io = 0; io < 2; io++) {
+    int max_pads = (io) ? NUM_DIO_PADS : NUM_MIO_PADS;
+    for (int i = 0; i < max_pads; i++) {
+      if (io == 1) {
+        // DIO
+        LOG_INFO("DIO [%d]: %x", i, kDioPads[i]);
+      } else {
+        // MIO
+        LOG_INFO("MIO [%d]: %x", i, kMioPads[i]);
+      }
+    }
+  }
+  LOG_INFO("END Chosen Retention Types");
+}
+
+bool lowpower_prep(dif_pwrmgr_t *pwrmgr, dif_pinmux_t *pinmux) {
+  bool result = false;
+  dif_pwrmgr_domain_config_t pwrmgr_domain_cfg;
+
+  LOG_INFO("Selecting PADs retention modes...");
+
+  draw_pinmux_ret(NUM_DIO_PADS, kDioPads);
+  draw_pinmux_ret(NUM_MIO_PADS, kMioPads);
+
+  print_chosen_values();
+
+  // Configure pwrmgr to deep powerdown
+  CHECK_DIF_OK(dif_pwrmgr_get_domain_config(pwrmgr, &pwrmgr_domain_cfg));
+  pwrmgr_domain_cfg =
+      pwrmgr_domain_cfg & (0xFF ^ kDifPwrmgrDomainOptionMainPowerInLowPower);
+
+  CHECK_DIF_OK(dif_pwrmgr_set_domain_config(pwrmgr, pwrmgr_domain_cfg,
+                                            kDifToggleEnabled));
+  CHECK_DIF_OK(dif_pwrmgr_low_power_set_enabled(pwrmgr, kDifToggleEnabled,
+                                                kDifToggleEnabled));
+
+  wait_for_interrupt();  // Entering deep power down
+
+  return result;
+}
+
+bool test_main(void) {
+  dif_pwrmgr_t pwrmgr;
+  dif_pinmux_t pinmux;
+
+  bool result = false;
+
+  CHECK_DIF_OK(dif_pwrmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+
+  if (pwrmgr_testutils_is_wakeup_reason(&pwrmgr, 0)) {
+    result = lowpower_prep(&pwrmgr, &pinmux);
+
+  } else if (pwrmgr_testutils_is_wakeup_reason(
+                 &pwrmgr, kDifPwrmgrWakeupRequestSourceThree)) {
+    // TODO: change PINMUX wakeup, not pin detector
+    /**
+     *  Usually this part won't be hit. UVM testbench checks the PAD output
```

Comment:
I would remove this condition as the UVM is taking care of it.

Reply:
That, still I didn't figure out how to finish in UVM tb successfully. In c code, SW returns `true` then, OTTF will take care of it. In UVM, I do not know yet. @cindychip @weicaiyang could you let me know?

Reply:
resolved in #14981